rewatch/src/project_context.rs

```diff
@@ -0,0 +1,288 @@
+        let path = helpers::get_abs_path(path);
+        let (current_config, current_rescript_json) = packages::read_config(&path)
+            .map_err(|_| anyhow!("Could not read rescript.json at {}", path.to_string_lossy()))?;
+        let nearest_parent_config_path = match path.parent() {
+            None => Err(anyhow!(
+                "The current path \"{}\" does not have a parent folder",
+                path.to_string_lossy()
+            )),
+            Some(parent) => Ok(helpers::get_nearest_config(parent)),
+        }?;
+        let context = match nearest_parent_config_path {
+            None => monorepo_or_single_project(&path, current_rescript_json, current_config),
+            Some(parent_config_path) => {
+                match packages::read_config(parent_config_path.as_path()) {
+                    Err(e) => Err(anyhow!(
+                        "Could not read the parent config at {}: {}",
+                        parent_config_path.to_string_lossy(),
+                        e
+                    )),
+                    Ok((workspace_config, workspace_rescript_json)) => {
+                        let is_current_config_listed_in_workspace = workspace_config
+                            .dependencies
+                            .to_owned()
+                            .unwrap_or_default()
+                            .iter()
+                            .any(|dep| dep == &current_config.name)
+                            || workspace_config
+                                .dev_dependencies
+                                .to_owned()
+                                .unwrap_or_default()
+                                .iter()
+                                .any(|dep| dep == &current_config.name);
+
+                        if is_current_config_listed_in_workspace {
+                            // There is a parent rescript.json, and it has a reference to the current package.
+                            Ok(ProjectContext::MonorepoPackage {
+                                config: current_config,
+                                path: current_rescript_json,
+                                parent_config: workspace_config,
+                                parent_path: workspace_rescript_json.clone(),
+                            })
+                        } else {
+                            // There is a parent rescript.json, but it has no reference to the current package.
+                            // However, the current package could still be a monorepo root!
+                            monorepo_or_single_project(&path, current_rescript_json, current_config)
+                        }
+                    }
+                }
+            }
+        };
+        context.iter().for_each(|pc| {
+            debug!("Created project context {:#?} for \"{}\"", pc, path.display());
+        });
+        context
+    }
+
+    pub fn get_root_config(&self) -> &Config {
```

Comment:
Breaking up the `enum` into a `struct` for the shared fields would remove the need for these too, which is nice imo :)

Reply:
Yeah, check you're right actually. Misread - these would still be there :) - still think having the overlapping fields in a struct would model the real world better though

Reply:
@rolandpeelen I've refactored based on the feedback,

now I have

```rs
#[allow(clippy::large_enum_variant)]
pub enum ProjectContext {
    /// Single project - no workspace relationship
    SingleProject { config: Config },
    /// Monorepo root - contains local dependencies (symlinked in node_modules)
    MonorepoRoot {
        config: Config,
        local_dependencies: AHashSet<String>, // names of local deps
        local_dev_dependencies: AHashSet<String>,
    },
    /// Package within a monorepo - has a parent workspace
    MonorepoPackage { config: Config, parent_config: Config },
}
```

How would change that? They all share their own `Config`. How to continue?

Reply:
I would personally still break it out to a `struct` + `enum`, all have the same `config` object at the root, and only when you have a monorepo, there are additional fields. 

So a project always has a config, but it may have a parent config (in the case of a package within a monorepo), or may have local dependencies (in case it's the root package) - but have no strong opinion.

The functions could still be there, because the 'root' would change in the case of a monorepo-package, but you'd for instance won't need the `get_current_rescript_config` - it would just be `x.config`, which makes a lot more sense to me than having an abstraction over a field that exists for all cases.